### ISSUE TITLE
Product unavailable error

### ISSUE THREAD

**Issue by mihai-moldovan** [Shopify/mobile-buy-sdk-ios#102]
Our shop contains exactly one variant for each product and one item for each variant. The user has the ability to view the cart and modify it (add/remove  products).

Scenario 1: 
The user goes for checkout. 
The app creates an empty BUYCart(), then adds the product variants to it, then creates the BUYCheckout with the cart and calls BUYClient.createCheckuot(). 
Then the user decides that she wants to modify the cart so she goes back, adds another product the process starts again.
The app has to recreate the cart since I cannot find a way to modify an existing cart, but the original products are already reserved so now createCheckout gives back the error: checkout-line_items-quantity-not_enought_in_stock but doesn't specify which product id gives the error.

Scenario 2:
For scneario 1, I found the method BUYClient.removeProductReservationsFromCheckout so let's say I call it when the user modifies the checkout so I can start from scratch. But imagine this one: the user goes for checkout, then she sends the app in background for whatever reason and the app gets killed. Coming back I need to recreate the cart and checkout, but again the products are already reserved.

Scenario 3:
Two users go for checkout pretty much in the same time, both having the same product in their cart. This is possible, because when getting the product details they show as available, although they are reserved. 
Again in the checkout process we get the same error but we cannot tell the user which product is already reserved by another user.

Bottom line: 
I'm either doing something wrong and there's a different way to implement this or we may need some of the following:

- a way to modify the products in an existing  BUYCart 

- meaningful errors for reserved products so that we can show the user something useful and also for any other error. Or a list of all possible error codes so we can treat them all inside the application.

- a way to serialise/deserialise the BUYCart object so we can do a proper cleanup when starting over (I
think we can maybe do this using the cartToken. Please confirm)

**Comment by davidmuzi** [comment]
The default reservation time is 5 minutes, but can be set to any value you like.  Even when the app goes into the background the reservation will expire, allowing the products to be added to someone else's checkout.  Alternatively, you could listen to `UIApplicationDidEnterBackgroundNotification`, ask for background time using `beginBackgroundTaskWithExpirationHandler` and remove the product reservations.

The error should contain the index of the item that there is no inventory for.  If this is not the case, please provide a sample response and we will look into it.

We are exploring ways to better represent errors, but unfortunately there is nothing we can share at this time.  There is no list of all possible errors.

You could modify the BUYCart implementation to persist it locally, and recreate the checkout from it between launches.

Hope this helps

**Comment by mihai-moldovan** [comment]
Thanks for the info.  So, if I understood correctly, when I get an error like this:

> {
>     errors =     {
>         checkout =         {
>             email =             (
>             );
>             "line_items" =             (
>                                 {
>                     quantity =                     (
>                                                 {
>                             code = "not_enough_in_stock";
>                             message = "Not enough items available. Only 0 left.";
>                             options =                             {
>                                 remaining = 0;
>                             };
>                         }
>                     );
>                 }
>             );
>             "reservation_time" =             (
>             );
>             "source_name" =             (
>             );
>         };
>     };
> }
line_items is an array and will contain a dictionary with quantity key only for the reserved products. Is this correct?

**Comment by davidmuzi** [comment]
Correct.  The index of the error corresponds to the index of the line item in the checkout.  So you may see <null> errors in there too.

**Comment by mihai-moldovan** [comment]
Can you provide an example of parsing the error in swift? 
Is there any way I can get a json of the error? 
This way I can just parse it using SwiftyJSON. Otherwise the parsing code gets very complicated since I have to unwrap and cast each value in the dictionary/array.

**Comment by davidmuzi** [comment]
The payload is in JSON, and can be readily parsed using a JSON serializer.  The errors are unique to each endpoint, and type, so unfortunately, there is no one solution for parsing them throughout the app.  We are working to improve this in a future update.

Will close this ticket, as its not an issue in the SDK

**Comment by hitendradeveloper** [comment]
https://cloud.githubusercontent.com/assets/9197369/23458294/acb2d9de-fea1-11e6-99cc-31dc75a4f5d1.png

inventory error response:
https://cloud.githubusercontent.com/assets/9197369/23458294/acb2d9de-fea1-11e6-99cc-31dc75a4f5d1.png



Shopify returns the error for variant in the same order as they are added in the BuyCart as lineItems, so you can map you cart products[array] with the inventory errors[array], and you can show the inventory error in the relative tableviewcell to guide the user to modify the quantity of the products
